pkg/plugins/runtime/k8s/controllers/gatewayapi/attachment/attachment.go

```diff
@@ -0,0 +1,181 @@
+package attachment
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	kube_apierrs "k8s.io/apimachinery/pkg/api/errors"
+	kube_apimeta "k8s.io/apimachinery/pkg/api/meta"
+	kube_meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kube_labels "k8s.io/apimachinery/pkg/labels"
+	kube_types "k8s.io/apimachinery/pkg/types"
+	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayapi "sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/controllers/gatewayapi/common"
+)
+
+type Attachment int
+
+const (
+	// Unknown means we shouldn't report on whether it's attached or not
+	Unknown Attachment = iota
+	// NotPermitted means the route isn't allowed to attach to the ref
+	NotPermitted
+	// Invalid means the route points to a nonexistent parent
+	// or is otherwise invalid
+	Invalid
+	// Attached means we can successfully attached
+	Attached
+)
+
+func listenerReady(
+	statuses []gatewayapi.ListenerStatus,
+	name gatewayapi.SectionName,
+) bool {
+	for _, status := range statuses {
+		if status.Name != name {
+			continue
+		}
+
+		return kube_apimeta.IsStatusConditionTrue(status.Conditions, string(gatewayapi.ListenerConditionReady))
+	}
+
+	return false
+}
+
+// findRouteListenerAttachment reports whether this ref is allowed to attach to
+// the Gateway or specific Listener.
+func findRouteListenerAttachment(
+	gateway *gatewayapi.Gateway,
+	routeNs kube_client.Object,
+	refSectionName *gatewayapi.SectionName,
+) (Attachment, error) {
+	// Build a map of whether attaching to each listener is possible
+	listeners := map[gatewayapi.SectionName]Attachment{}
+
+	for _, l := range gateway.Spec.Listeners {
+		ns := l.AllowedRoutes.Namespaces
+
+		// TODO
+		// The gateway controller ensures Kinds is either empty or contains
+		// HTTPRoute, so we can ignore it here.
+
+		switch *ns.From {
+		case gatewayapi.NamespacesFromSelector:
+			// TODO, the gateway controller/webhook should verify this
+			selector, err := kube_meta.LabelSelectorAsSelector(ns.Selector)
+			if err != nil {
+				return Unknown, errors.Wrap(err, "internal error: couldn't convert to selector")
+			}
+
+			if !selector.Matches(kube_labels.Set(routeNs.GetLabels())) {
+				listeners[l.Name] = NotPermitted
+				continue
+			}
+		case gatewayapi.NamespacesFromSame:
+			if gateway.Namespace != routeNs.GetName() {
+				listeners[l.Name] = NotPermitted
+				continue
+			}
+		case gatewayapi.NamespacesFromAll:
+		}
+
+		if listenerReady(gateway.Status.Listeners, l.Name) {
+			listeners[l.Name] = Attached
+		} else {
+			listeners[l.Name] = Unknown
+		}
+	}
+
```

Comment:
I got a bit confused by the listener machinery below. The way that helped me was to realize that if `refSectionName` is nil, the reference must be allowed for all listeners, but if it is not nil, it only needs to be allowed for the named listener.

This is arguably ambiguous in the spec, and I think that the last loop that assumes one listener is enough is not correct.

Reply:
I took it from [upstream](https://github.com/kubernetes-sigs/gateway-api/blob/c6729305f8baf14c0660a5ef5bce49c2aaf35759/apis/v1alpha2/shared_types.go#L74-L81):

> 	// When unspecified (empty string), this will reference the entire resource.
	// For the purpose of status, an attachment is considered successful if at
	// least one section in the parent resource accepts it. For example, Gateway
	// listeners can restrict which Routes can attach to them by Route kind,
	// namespace, or hostname. If 1 of 2 Gateway listeners accept attachment from
	// the referencing Route, the Route MUST be considered successfully
	// attached. If no Gateway listeners accept attachment from this Route, the
	// Route MUST be considered detached from the Gateway.

Reply:
So the rule is that any listener can accept the route? That's something I found confusing about this PR in general - it's testing whether a route can be attached to the gateway, but a more direct reading of the spec is whether the route can be attached to the listener. That perspective assumes that you are starting with a listener, finding the set of routes and then testing them all. This code is inverted (it starts with a route) and it takes some time to thing it through to convince yourself that it works.

Reply:
> That's something I found confusing about this PR in general - it's testing whether a route can be attached to the gateway, but a more direct reading of the spec is whether the route can be attached to the listener.

I don't think it's quite testing if a route can attach to a gateway, it's testing whether the route can attach to a parent. A parent reference can be either a specific listener or an entire gateway.

> That perspective assumes that you are starting with a listener

This isn't related to whether it's listener or gateway though, it's about whether you conceptually iterate over gateways (and thus listeners) to find matching routes or do you go through each route and see whether they can attach to their listed parents.

You're right, the spec wording seems to imply the _Gateway controller_ sets the `HTTPRoute` `RouteParentStatus` whereas here we use the `HTTPRoute` controller to set the status.
The behavior is the same though I think, because we reconcile `HTTPRoute`s whenever a `Gateway` changes (and thus iterate over all `HTTPRoute`s either way). I think either way one controller ends up manipulating the status of another. Now that I think about it, the `HTTPRoute` controller should also be incrementing/decrementing the `GatewayStatus.AttachedRoutes` if its `Accepted` status changes.

> So the rule is that any listener can accept the route? 

Yes, I think that's the semantics of not setting `SectionName` in an `HTTPRoute` `ParentRef`. Any listener on the Gateway that permits the HTTPRoute should attach the route and as long as some listener allows it, that route's Status says it's attached to its parent.


Does that all make sense?

Reply:
Resolved for now.